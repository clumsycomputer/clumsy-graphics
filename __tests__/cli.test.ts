import ChildProcess from 'child_process'
import FileSystem from 'fs'
import { test, expect } from '@playwright/test'

test.describe('graphics-renderer', () => {
  const testAnimationModulePath = './__tests__/test-project/Foo.animation.tsx'
  test.beforeAll(() => {
    FileSystem.rmSync('./test-temp', {
      recursive: true,
      force: true,
    })
    FileSystem.mkdirSync('./test-temp')
  })
  test.afterAll(() => {
    FileSystem.rmSync('./test-temp', {
      recursive: true,
      force: true,
    })
  })
  test.describe('renderAnimation', () => {
    test('renderAnimation outputs expected "foo.mp4"', () => {
      ChildProcess.execSync(
        `graphics-renderer renderAnimation --animationModulePath=${testAnimationModulePath} --animationMp4OutputPath=./test-temp/foo.mp4 --numberOfFrameRendererWorkers=7`
      )
      expect(
        ChildProcess.spawnSync(
          'ffprobe',
          ['-show_frames', './test-temp/foo.mp4'],
          { stdio: 'pipe' }
        ).stdout.toString()
      ).toMatchSnapshot('animationMp4FramesSnapshot')
    })
  })
  test.describe('renderAnimationFrame', () => {
    test('renderAnimationFrame outputs expected "foo_0.svg"', () => {
      ChildProcess.execSync(
        `graphics-renderer renderAnimationFrame --animationModulePath=${testAnimationModulePath} --frameFileOutputPath=./test-temp/foo_0.svg --frameIndex=0`
      )
      expect(
        FileSystem.readFileSync('./test-temp/foo_0.svg')
          .toString()
          .replace(/sodipodi:docname="(.)+\.svg"/m, '')
      ).toMatchSnapshot('frameSvgSnapshot')
    })
    test('renderAnimationFrame outputs expected "foo_0.png"', () => {
      ChildProcess.execSync(
        `graphics-renderer renderAnimationFrame --animationModulePath=${testAnimationModulePath} --frameFileOutputPath=./test-temp/foo_0.png --frameIndex=0`
      )
      expect(
        ChildProcess.spawnSync(
          'ffprobe',
          ['-show_frames', './test-temp/foo_0.png'],
          { stdio: 'pipe' }
        ).stdout.toString()
      ).toMatchSnapshot('framePngFramesSnapshot')
    })
  })
  test.describe('convertAnimationToGif', () => {
    test('convertAnimationToGif outputs expected "foo.gif"', () => {
      ChildProcess.execSync(
        `graphics-renderer renderAnimation --animationModulePath=${testAnimationModulePath} --animationMp4OutputPath=./test-temp/foo.mp4`
      )
      ChildProcess.execSync(
        'graphics-renderer convertAnimationToGif --animationMp4SourcePath=./test-temp/foo.mp4 --animationGifOutputPath=./test-temp/foo.gif --gifAspectRatioWidth=512'
      )
      expect(
        ChildProcess.spawnSync(
          'ffprobe',
          ['-show_frames', './test-temp/foo.gif'],
          { stdio: 'pipe' }
        ).stdout.toString()
      ).toMatchSnapshot('animationGifFramesSnapshot')
    })
  })
  test.describe('startAnimationDevelopment', () => {
    let spawnedDevelopmentProcess: ChildProcess.ChildProcess | null = null
    test.beforeAll(async () => {
      spawnedDevelopmentProcess = ChildProcess.spawn('graphics-renderer', [
        'startDevelopment',
        `--animationModulePath=${testAnimationModulePath}`,
      ])
      await new Promise<void>((resolve) => {
        setTimeout(() => {
          resolve()
        }, 1000)
      })
    })
    test.afterAll(() => {
      spawnedDevelopmentProcess!.kill('SIGINT')
    })
    test('startAnimationDevelopment renders and serves animation asset', async ({
      page,
    }) => {
      await page.goto('localhost:3000/?assetType=mp4')
      const assetResponse = await page.waitForResponse('**/asset/0.mp4')
      expect(await assetResponse.headerValues('Content-Type')).toEqual([
        'video/mp4',
      ])
    })
    test('startAnimationDevelopment renders and serves frame asset', async ({
      page,
    }) => {
      await page.goto('localhost:3000/?assetType=png&frameIndex=0')
      const assetResponse = await page.waitForResponse('**/asset/0_0.png')
      expect(await assetResponse.headerValues('Content-Type')).toEqual([
        'image/png',
      ])
    })
  })
})
