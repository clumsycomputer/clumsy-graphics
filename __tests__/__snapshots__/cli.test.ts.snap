// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`graphics-renderer convertAnimationToGif convertAnimationToGif outputs expected "foo.gif" 1`] = `
"[FRAME]
media_type=video
stream_index=0
key_frame=1
pkt_pts=0
pkt_pts_time=0.000000
pkt_dts=0
pkt_dts_time=0.000000
best_effort_timestamp=0
best_effort_timestamp_time=0.000000
pkt_duration=20
pkt_duration_time=0.200000
pkt_pos=0
pkt_size=3219
width=512
height=512
pix_fmt=bgra
sample_aspect_ratio=1:1
pict_type=I
coded_picture_number=0
display_picture_number=0
interlaced_frame=0
top_field_first=0
repeat_pict=0
color_range=unknown
color_space=unknown
color_primaries=unknown
color_transfer=unknown
chroma_location=unspecified
[/FRAME]
[FRAME]
media_type=video
stream_index=0
key_frame=0
pkt_pts=20
pkt_pts_time=0.200000
pkt_dts=20
pkt_dts_time=0.200000
best_effort_timestamp=20
best_effort_timestamp_time=0.200000
pkt_duration=20
pkt_duration_time=0.200000
pkt_pos=3219
pkt_size=1167
width=512
height=512
pix_fmt=bgra
sample_aspect_ratio=1:1
pict_type=P
coded_picture_number=0
display_picture_number=0
interlaced_frame=0
top_field_first=0
repeat_pict=0
color_range=unknown
color_space=unknown
color_primaries=unknown
color_transfer=unknown
chroma_location=unspecified
[/FRAME]
[FRAME]
media_type=video
stream_index=0
key_frame=0
pkt_pts=40
pkt_pts_time=0.400000
pkt_dts=40
pkt_dts_time=0.400000
best_effort_timestamp=40
best_effort_timestamp_time=0.400000
pkt_duration=20
pkt_duration_time=0.200000
pkt_pos=4386
pkt_size=1191
width=512
height=512
pix_fmt=bgra
sample_aspect_ratio=1:1
pict_type=P
coded_picture_number=0
display_picture_number=0
interlaced_frame=0
top_field_first=0
repeat_pict=0
color_range=unknown
color_space=unknown
color_primaries=unknown
color_transfer=unknown
chroma_location=unspecified
[/FRAME]
[FRAME]
media_type=video
stream_index=0
key_frame=0
pkt_pts=60
pkt_pts_time=0.600000
pkt_dts=60
pkt_dts_time=0.600000
best_effort_timestamp=60
best_effort_timestamp_time=0.600000
pkt_duration=20
pkt_duration_time=0.200000
pkt_pos=5577
pkt_size=1218
width=512
height=512
pix_fmt=bgra
sample_aspect_ratio=1:1
pict_type=P
coded_picture_number=0
display_picture_number=0
interlaced_frame=0
top_field_first=0
repeat_pict=0
color_range=unknown
color_space=unknown
color_primaries=unknown
color_transfer=unknown
chroma_location=unspecified
[/FRAME]
[FRAME]
media_type=video
stream_index=0
key_frame=0
pkt_pts=80
pkt_pts_time=0.800000
pkt_dts=80
pkt_dts_time=0.800000
best_effort_timestamp=80
best_effort_timestamp_time=0.800000
pkt_duration=20
pkt_duration_time=0.200000
pkt_pos=6795
pkt_size=1171
width=512
height=512
pix_fmt=bgra
sample_aspect_ratio=1:1
pict_type=P
coded_picture_number=0
display_picture_number=0
interlaced_frame=0
top_field_first=0
repeat_pict=0
color_range=unknown
color_space=unknown
color_primaries=unknown
color_transfer=unknown
chroma_location=unspecified
[/FRAME]
[FRAME]
media_type=video
stream_index=0
key_frame=0
pkt_pts=100
pkt_pts_time=1.000000
pkt_dts=100
pkt_dts_time=1.000000
best_effort_timestamp=100
best_effort_timestamp_time=1.000000
pkt_duration=20
pkt_duration_time=0.200000
pkt_pos=7966
pkt_size=1058
width=512
height=512
pix_fmt=bgra
sample_aspect_ratio=1:1
pict_type=P
coded_picture_number=0
display_picture_number=0
interlaced_frame=0
top_field_first=0
repeat_pict=0
color_range=unknown
color_space=unknown
color_primaries=unknown
color_transfer=unknown
chroma_location=unspecified
[/FRAME]
[FRAME]
media_type=video
stream_index=0
key_frame=0
pkt_pts=120
pkt_pts_time=1.200000
pkt_dts=120
pkt_dts_time=1.200000
best_effort_timestamp=120
best_effort_timestamp_time=1.200000
pkt_duration=20
pkt_duration_time=0.200000
pkt_pos=9024
pkt_size=1131
width=512
height=512
pix_fmt=bgra
sample_aspect_ratio=1:1
pict_type=P
coded_picture_number=0
display_picture_number=0
interlaced_frame=0
top_field_first=0
repeat_pict=0
color_range=unknown
color_space=unknown
color_primaries=unknown
color_transfer=unknown
chroma_location=unspecified
[/FRAME]
[FRAME]
media_type=video
stream_index=0
key_frame=0
pkt_pts=140
pkt_pts_time=1.400000
pkt_dts=140
pkt_dts_time=1.400000
best_effort_timestamp=140
best_effort_timestamp_time=1.400000
pkt_duration=20
pkt_duration_time=0.200000
pkt_pos=10155
pkt_size=1248
width=512
height=512
pix_fmt=bgra
sample_aspect_ratio=1:1
pict_type=P
coded_picture_number=0
display_picture_number=0
interlaced_frame=0
top_field_first=0
repeat_pict=0
color_range=unknown
color_space=unknown
color_primaries=unknown
color_transfer=unknown
chroma_location=unspecified
[/FRAME]
[FRAME]
media_type=video
stream_index=0
key_frame=0
pkt_pts=160
pkt_pts_time=1.600000
pkt_dts=160
pkt_dts_time=1.600000
best_effort_timestamp=160
best_effort_timestamp_time=1.600000
pkt_duration=20
pkt_duration_time=0.200000
pkt_pos=11403
pkt_size=1207
width=512
height=512
pix_fmt=bgra
sample_aspect_ratio=1:1
pict_type=P
coded_picture_number=0
display_picture_number=0
interlaced_frame=0
top_field_first=0
repeat_pict=0
color_range=unknown
color_space=unknown
color_primaries=unknown
color_transfer=unknown
chroma_location=unspecified
[/FRAME]
[FRAME]
media_type=video
stream_index=0
key_frame=0
pkt_pts=180
pkt_pts_time=1.800000
pkt_dts=180
pkt_dts_time=1.800000
best_effort_timestamp=180
best_effort_timestamp_time=1.800000
pkt_duration=20
pkt_duration_time=0.200000
pkt_pos=12610
pkt_size=1219
width=512
height=512
pix_fmt=bgra
sample_aspect_ratio=1:1
pict_type=P
coded_picture_number=0
display_picture_number=0
interlaced_frame=0
top_field_first=0
repeat_pict=0
color_range=unknown
color_space=unknown
color_primaries=unknown
color_transfer=unknown
chroma_location=unspecified
[/FRAME]
"
`;

exports[`graphics-renderer renderAnimation renderAnimation outputs expected "foo.mp4" 1`] = `
"[FRAME]
media_type=video
stream_index=0
key_frame=1
pkt_pts=0
pkt_pts_time=0.000000
pkt_dts=0
pkt_dts_time=0.000000
best_effort_timestamp=0
best_effort_timestamp_time=0.000000
pkt_duration=2048
pkt_duration_time=0.200000
pkt_pos=48
pkt_size=2024
width=512
height=512
pix_fmt=yuv420p
sample_aspect_ratio=1:1
pict_type=I
coded_picture_number=0
display_picture_number=0
interlaced_frame=0
top_field_first=0
repeat_pict=0
color_range=unknown
color_space=unknown
color_primaries=unknown
color_transfer=unknown
chroma_location=left
[SIDE_DATA]
side_data_type=H.26[45] User Data Unregistered SEI message
[/SIDE_DATA]
[/FRAME]
[FRAME]
media_type=video
stream_index=0
key_frame=0
pkt_pts=2048
pkt_pts_time=0.200000
pkt_dts=2048
pkt_dts_time=0.200000
best_effort_timestamp=2048
best_effort_timestamp_time=0.200000
pkt_duration=2048
pkt_duration_time=0.200000
pkt_pos=2072
pkt_size=1303
width=512
height=512
pix_fmt=yuv420p
sample_aspect_ratio=1:1
pict_type=P
coded_picture_number=1
display_picture_number=0
interlaced_frame=0
top_field_first=0
repeat_pict=0
color_range=unknown
color_space=unknown
color_primaries=unknown
color_transfer=unknown
chroma_location=left
[/FRAME]
[FRAME]
media_type=video
stream_index=0
key_frame=0
pkt_pts=4096
pkt_pts_time=0.400000
pkt_dts=4096
pkt_dts_time=0.400000
best_effort_timestamp=4096
best_effort_timestamp_time=0.400000
pkt_duration=2048
pkt_duration_time=0.200000
pkt_pos=3375
pkt_size=1303
width=512
height=512
pix_fmt=yuv420p
sample_aspect_ratio=1:1
pict_type=P
coded_picture_number=2
display_picture_number=0
interlaced_frame=0
top_field_first=0
repeat_pict=0
color_range=unknown
color_space=unknown
color_primaries=unknown
color_transfer=unknown
chroma_location=left
[/FRAME]
[FRAME]
media_type=video
stream_index=0
key_frame=0
pkt_pts=6144
pkt_pts_time=0.600000
pkt_dts=6144
pkt_dts_time=0.600000
best_effort_timestamp=6144
best_effort_timestamp_time=0.600000
pkt_duration=2048
pkt_duration_time=0.200000
pkt_pos=4678
pkt_size=851
width=512
height=512
pix_fmt=yuv420p
sample_aspect_ratio=1:1
pict_type=P
coded_picture_number=3
display_picture_number=0
interlaced_frame=0
top_field_first=0
repeat_pict=0
color_range=unknown
color_space=unknown
color_primaries=unknown
color_transfer=unknown
chroma_location=left
[/FRAME]
[FRAME]
media_type=video
stream_index=0
key_frame=0
pkt_pts=8192
pkt_pts_time=0.800000
pkt_dts=8192
pkt_dts_time=0.800000
best_effort_timestamp=8192
best_effort_timestamp_time=0.800000
pkt_duration=2048
pkt_duration_time=0.200000
pkt_pos=5529
pkt_size=1000
width=512
height=512
pix_fmt=yuv420p
sample_aspect_ratio=1:1
pict_type=P
coded_picture_number=4
display_picture_number=0
interlaced_frame=0
top_field_first=0
repeat_pict=0
color_range=unknown
color_space=unknown
color_primaries=unknown
color_transfer=unknown
chroma_location=left
[/FRAME]
[FRAME]
media_type=video
stream_index=0
key_frame=0
pkt_pts=10240
pkt_pts_time=1.000000
pkt_dts=10240
pkt_dts_time=1.000000
best_effort_timestamp=10240
best_effort_timestamp_time=1.000000
pkt_duration=2048
pkt_duration_time=0.200000
pkt_pos=6529
pkt_size=1301
width=512
height=512
pix_fmt=yuv420p
sample_aspect_ratio=1:1
pict_type=P
coded_picture_number=5
display_picture_number=0
interlaced_frame=0
top_field_first=0
repeat_pict=0
color_range=unknown
color_space=unknown
color_primaries=unknown
color_transfer=unknown
chroma_location=left
[/FRAME]
[FRAME]
media_type=video
stream_index=0
key_frame=0
pkt_pts=12288
pkt_pts_time=1.200000
pkt_dts=12288
pkt_dts_time=1.200000
best_effort_timestamp=12288
best_effort_timestamp_time=1.200000
pkt_duration=2048
pkt_duration_time=0.200000
pkt_pos=7830
pkt_size=1267
width=512
height=512
pix_fmt=yuv420p
sample_aspect_ratio=1:1
pict_type=P
coded_picture_number=6
display_picture_number=0
interlaced_frame=0
top_field_first=0
repeat_pict=0
color_range=unknown
color_space=unknown
color_primaries=unknown
color_transfer=unknown
chroma_location=left
[/FRAME]
[FRAME]
media_type=video
stream_index=0
key_frame=0
pkt_pts=14336
pkt_pts_time=1.400000
pkt_dts=14336
pkt_dts_time=1.400000
best_effort_timestamp=14336
best_effort_timestamp_time=1.400000
pkt_duration=2048
pkt_duration_time=0.200000
pkt_pos=9097
pkt_size=1610
width=512
height=512
pix_fmt=yuv420p
sample_aspect_ratio=1:1
pict_type=P
coded_picture_number=7
display_picture_number=0
interlaced_frame=0
top_field_first=0
repeat_pict=0
color_range=unknown
color_space=unknown
color_primaries=unknown
color_transfer=unknown
chroma_location=left
[/FRAME]
[FRAME]
media_type=video
stream_index=0
key_frame=0
pkt_pts=16384
pkt_pts_time=1.600000
pkt_dts=N/A
pkt_dts_time=N/A
best_effort_timestamp=16384
best_effort_timestamp_time=1.600000
pkt_duration=2048
pkt_duration_time=0.200000
pkt_pos=11960
pkt_size=797
width=512
height=512
pix_fmt=yuv420p
sample_aspect_ratio=1:1
pict_type=B
coded_picture_number=9
display_picture_number=0
interlaced_frame=0
top_field_first=0
repeat_pict=0
color_range=unknown
color_space=unknown
color_primaries=unknown
color_transfer=unknown
chroma_location=left
[/FRAME]
[FRAME]
media_type=video
stream_index=0
key_frame=0
pkt_pts=18432
pkt_pts_time=1.800000
pkt_dts=N/A
pkt_dts_time=N/A
best_effort_timestamp=18432
best_effort_timestamp_time=1.800000
pkt_duration=2048
pkt_duration_time=0.200000
pkt_pos=10707
pkt_size=1253
width=512
height=512
pix_fmt=yuv420p
sample_aspect_ratio=1:1
pict_type=P
coded_picture_number=8
display_picture_number=0
interlaced_frame=0
top_field_first=0
repeat_pict=0
color_range=unknown
color_space=unknown
color_primaries=unknown
color_transfer=unknown
chroma_location=left
[/FRAME]
"
`;

exports[`graphics-renderer renderAnimationFrame renderAnimationFrame outputs expected "foo_0.png" 1`] = `
"[FRAME]
media_type=video
stream_index=0
key_frame=1
pkt_pts=0
pkt_pts_time=0.000000
pkt_dts=0
pkt_dts_time=0.000000
best_effort_timestamp=0
best_effort_timestamp_time=0.000000
pkt_duration=1
pkt_duration_time=0.040000
pkt_pos=0
pkt_size=3395
width=512
height=512
pix_fmt=rgba
sample_aspect_ratio=1:1
pict_type=I
coded_picture_number=0
display_picture_number=0
interlaced_frame=0
top_field_first=0
repeat_pict=0
color_range=pc
color_space=unknown
color_primaries=unknown
color_transfer=unknown
chroma_location=unspecified
TAG:Software=www.inkscape.org
[/FRAME]
"
`;

exports[`graphics-renderer renderAnimationFrame renderAnimationFrame outputs expected "foo_0.svg" 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\" standalone=\\"no\\"?>
<svg
   viewBox=\\"0 0 100 100\\"
   version=\\"1.1\\"
   id=\\"svg6\\"
   
   xmlns:inkscape=\\"http://www.inkscape.org/namespaces/inkscape\\"
   xmlns:sodipodi=\\"http://sodipodi.sourceforge.net/DTD/sodipodi-0.dtd\\"
   xmlns=\\"http://www.w3.org/2000/svg\\"
   xmlns:svg=\\"http://www.w3.org/2000/svg\\">
  <defs
     id=\\"defs10\\" />
  <sodipodi:namedview
     id=\\"namedview8\\"
     pagecolor=\\"#ffffff\\"
     bordercolor=\\"#666666\\"
     borderopacity=\\"1.0\\"
     inkscape:pageshadow=\\"2\\"
     inkscape:pageopacity=\\"0.0\\"
     inkscape:pagecheckerboard=\\"0\\" />
  <rect
     x=\\"0\\"
     y=\\"0\\"
     width=\\"100\\"
     height=\\"100\\"
     fill=\\"black\\"
     id=\\"rect2\\" />
  <circle
     cx=\\"65\\"
     cy=\\"50\\"
     r=\\"5\\"
     fill=\\"rgb(0,128,64)\\"
     id=\\"circle4\\" />
</svg>
"
`;
